### ISSUE TITLE
Realm Object empty

### ISSUE THREAD

**Issue by iwijk** [evermeer/EVReflection#227]
I have a issue when converting a Realm object to a jsonString. I searched online but couldn't find a clear answer. I have a guess what might be the issue...

I have a user Class:
```swift
import RealmSwift
import EVReflection

public class User: Object, EVReflectable, IModel {
    dynamic var id: String = UUID().uuidString
    
    dynamic var createdAt: String = ""
    dynamic var lastUpdate: String = ""
    
    dynamic var firstName: String = ""
    dynamic var lastName: String = ""
    
    dynamic var email: String = ""
}
```

When I try to load the user from my realm database like so:
```swift
let users = realm.objects(User.self).filter("id = %@", userId)
let user = users.first
```

I can access the object without a problem. I see all the data that are stored in the object.

If I want to convert the data to a JSON string and output it I get an empty json object:
```swift
print(user.toJsonString())
```
```json
{
   "id":"64EA49AA-333C-4C76-B92C-464EFE1D9E10",
   "createdAt":"",
   "lastUpdate":"",
   "firstName":"",
   "lastName":"",
   "email":""
}
```

Is this a bug or a Realm related behaviour. Because the only explanation I can find that explains this kind of behaviour is this issue: [Github Realm Issue](https://github.com/realm/realm-cocoa/issues/2777#issuecomment-152984130)

**Comment by evermeer** [comment]
That sounds like the cause of your issue. I guess I have to find out if it's possible to get the values from the object in any other way than using Swift Mirror. I had a similar issue with CoreData. I'm not very experienced with Realm, so any tips would be welcome.

**Comment by iwijk** [comment]
I do not have that much experience with Realm either, only used it in one project. But I do know that Realm makes use of so called "Dummy" objects. So in my understanding it does not access the data at the time it is "pulled" from the database, but at the time it the object is used in the code.

My guess is that by using the following technique: (If I saw it correctly in your code, the same methode you are using)
https://stackoverflow.com/a/32969324/793166

https://developer.apple.com/documentation/swift/mirror
"Mirrors are used by playgrounds and the debugger."

Because of this I think the only solution is to "manual/static" map the object to a dictionary.

**Comment by iwijk** [comment]
I did some more investigation:
"Note that although the LLDB script allows inspecting the contents of your Realm variables in Xcode’s UI, this doesn’t yet work for Swift. Instead, those variables will show incorrect data. You should instead use LLDB’s `po` command to inspect the contents of data stored in a Realm."
https://realm.io/docs/swift/latest/#debugging

"Print Description of “{variable}”. An equivalent to using `po` in the console or using the Print Description button in a tool tip."
https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/debugging_with_xcode/chapters/debugging_tools.html

I just tested the command `po` and I do see the correct values in my console. I think if we can figure out the technique used in to `po` command, we might have a solution.

**Comment by evermeer** [comment]
I'm not able to reproduce this. When I read a realmobject and print the json, then it's valid json.

See the unit test at:
https://github.com/evermeer/EVReflection/blob/master/UnitTests/RealmTests/RealmTests.swift#L94

You have to be aware that the current master is for Swift 4.

In your object definitions you specified they implemented EVReflectable. That's not needed. If you Import EVReflection, then Object is extended with EVReflectable.

**Comment by tprovoost** [comment]
Hi,

I happen to have the same problem, using the swift 3.2 branch of EVReflection and RealmSwift in 2.4. 

Basically, the methods `toDictionary()` and `toJsonString()` both have default initialized data instead of the actual data.

**Comment by tprovoost** [comment]
Actually created a new project in Swift 4, added only RealmSwift and EVReflection, both to last versions.
I copied your code from your test you linked above, and the json doesn't work: 

`object = {"age":0,"cars":[{"brand":"","name":null,"year":0},{"brand":"","name":null,"year":0}],"spouse":null,"name":""}`

Information is correctly displayed after: 
```
Jennifer
47
car.name = Outatime
car.brand = DeLorean
...
```

I think the reason is that the configuration in memory in your test actually makes the test pass, but in real conditions, it doesn't work. :(

**Comment by tprovoost** [comment]
Sorry for the "evening spam", but this is a nice workaround I found. Basically, this adds an extension to simply create a detached copy of an object:

https://github.com/realm/realm-cocoa/issues/3381#issuecomment-256243390

This solved the problem instantly. just by calling object.detached() and then `toJsonString()`.

**Comment by evermeer** [comment]
Thanks @tprovoost  that seems to be an easy fix for this issue. I will have a look at it this evening (in an app instead of a unit test).

**Comment by evermeer** [comment]
@tprovoost OK, I do have it working... but... It will crash if you have a recursive object reference like I have in my case (A person will have a spouse which is also a person that has the first person as its spouse)... So I have to find out a way to skip the detach if the object is already detached.

**Comment by tprovoost** [comment]
@evermeer Indeed, I could see that there is a problem with circular reference as you mentioned. But technically, I think that design-wise, circular references aren't so good in the first place. In your case, changing the model to the person's ID for instance would be a better approach.

**Comment by evermeer** [comment]
I have pushed the fix as version 5.0.3
The circular reference issue is not solved yet.

**Comment by tobyworks** [comment]
For me this problem occured in EVReflection 5.10.1
